lib/broadway_kafka/producer.ex

```diff
@@ -81,12 +81,32 @@ defmodule BroadwayKafka.Producer do
       throughput at the cost of more memory consumption.
 
     * `:max_wait_time` - Optional. Time in millisecond. Max number of milliseconds allowed for the broker to collect
-    `min_bytes' of messages in fetch response. Default is 1000ms.
+    `min_bytes` of messages in fetch response. Default is 1000ms.
 
   ## Client config options
 
   The available options that will be internally passed to `:brod.start_client/3`.
 
+    * `:client_id` - Optional. An atom that will be passed to `:brod` as the client id parameter for the connection. Example:
+    `client_id: :"\#{Node.self()} - \#{__MODULE__}"`. This generates the following connection log from our integration tests:
```

Comment:
There's a potential trap here: setting the same client_id for multiple producers (as you're doing in the example), would change the client:producer relation from 1:1 to 1:N. I guess a client could bottleneck, given enough producers?

But there's also a good side on this: being able to share a client between multiple Broadway topologies seems like a good idea depending on your architecture.

IMO, this should be documented somehow, wdyt?

Reply:
Actually, I'm not really sure if the current implementation won't crash if there's already a client with this name. 

If it does not, the example should crash if multiple producers are enabled for the topology.

Reply:
This does not change the old behaviour in anyway I think. The only difference is that before it created an alias of the producer module suffixed with `Client` and now it is your choice to customize this or not. If by "chance" you had a module with that alias it would result in the same issues you brought here (or non-issues as you verified that the client is idempotent).

We can, and perhaps should, enhance the docs with things like what you mention and some other "hidden gems" like defining a custom client. I believe this should be in a separate PR though.

Reply:
I don't think you understand what I pointed out. Try to folllow me:

- Your topology spawns N producers, they're named processes
- Each producer calls `:brodstart_client` with a name
- Without a client `id`, name is `my_name <> Client`, thus, `:brod.start_client` is called with many different names, you can see that when I start a Broadway topology with 5 producers:
```
%{
  client_config: [],
  fetch_config: %{},
  group_config: [offset_commit_policy: :commit_to_kafka_v2],
  group_id: "group_1",
  hosts: [localhost: 9092],
  offset_commit_on_ack: true,
  offset_reset_policy: :latest,
  receive_interval: 2000,
  reconnect_timeout: 1000,
  topics: ["my_topic", "test"]
}
MyBroadwayExample.Broadway.Producer_0.Client

16:43:58.673 [info]      :supervisor: {:local, :brod_sup}
    :started: [
  pid: #PID<0.310.0>,
  id: MyBroadwayExample.Broadway.Producer_0.Client,
  mfargs: {:brod_client, :start_link,
   [[localhost: 9092], MyBroadwayExample.Broadway.Producer_0.Client, []]},
  restart_type: {:permanent, 10},
  shutdown: 5000,
  child_type: :worker
]
%{
  client_config: [],
  fetch_config: %{},
  group_config: [offset_commit_policy: :commit_to_kafka_v2],
  group_id: "group_1",
  hosts: [localhost: 9092],
  offset_commit_on_ack: true,
  offset_reset_policy: :latest,
  receive_interval: 2000,
  reconnect_timeout: 1000,
  topics: ["my_topic", "test"]
}
MyBroadwayExample.Broadway.Producer_1.Client

16:43:58.699 [info]      :supervisor: {:local, :brod_sup}
    :started: [
  pid: #PID<0.316.0>,
  id: MyBroadwayExample.Broadway.Producer_1.Client,
  mfargs: {:brod_client, :start_link,
   [[localhost: 9092], MyBroadwayExample.Broadway.Producer_1.Client, []]},
  restart_type: {:permanent, 10},
  shutdown: 5000,
  child_type: :worker
]
%{
  client_config: [],
  fetch_config: %{},
  group_config: [offset_commit_policy: :commit_to_kafka_v2],
  group_id: "group_1",
  hosts: [localhost: 9092],
  offset_commit_on_ack: true,
  offset_reset_policy: :latest,
  receive_interval: 2000,
  reconnect_timeout: 1000,
  topics: ["my_topic", "test"]
}
MyBroadwayExample.Broadway.Producer_2.Client

16:43:58.700 [info]      :supervisor: {:local, :brod_sup}
    :started: [
  pid: #PID<0.320.0>,
  id: MyBroadwayExample.Broadway.Producer_2.Client,
  mfargs: {:brod_client, :start_link,
   [[localhost: 9092], MyBroadwayExample.Broadway.Producer_2.Client, []]},
  restart_type: {:permanent, 10},
  shutdown: 5000,
  child_type: :worker
]
%{
  client_config: [],
  fetch_config: %{},
  group_config: [offset_commit_policy: :commit_to_kafka_v2],
  group_id: "group_1",
  hosts: [localhost: 9092],
  offset_commit_on_ack: true,
  offset_reset_policy: :latest,
  receive_interval: 2000,
  reconnect_timeout: 1000,
  topics: ["my_topic", "test"]
}
MyBroadwayExample.Broadway.Producer_3.Client

16:43:58.700 [info]      :supervisor: {:local, :brod_sup}
    :started: [
  pid: #PID<0.324.0>,
  id: MyBroadwayExample.Broadway.Producer_3.Client,
  mfargs: {:brod_client, :start_link,
   [[localhost: 9092], MyBroadwayExample.Broadway.Producer_3.Client, []]},
  restart_type: {:permanent, 10},
  shutdown: 5000,
  child_type: :worker
]
%{
  client_config: [],
  fetch_config: %{},
  group_config: [offset_commit_policy: :commit_to_kafka_v2],
  group_id: "group_1",
  hosts: [localhost: 9092],
  offset_commit_on_ack: true,
  offset_reset_policy: :latest,
  receive_interval: 2000,
  reconnect_timeout: 1000,
  topics: ["my_topic", "test"]
}
MyBroadwayExample.Broadway.Producer_4.Client

16:43:58.701 [info]      :supervisor: {:local, :brod_sup}
    :started: [
  pid: #PID<0.328.0>,
  id: MyBroadwayExample.Broadway.Producer_4.Client,
  mfargs: {:brod_client, :start_link,
   [[localhost: 9092], MyBroadwayExample.Broadway.Producer_4.Client, []]},
  restart_type: {:permanent, 10},
  shutdown: 5000,
  child_type: :worker
]
```

- Now, if I set the `client_id`, all my 5 producers share the same brod client:
```
%{
  client_config: [client_id: :my_client],
  fetch_config: %{},
  group_config: [offset_commit_policy: :commit_to_kafka_v2],
  group_id: "group_1",
  hosts: [localhost: 9092],
  offset_commit_on_ack: true,
  offset_reset_policy: :latest,
  receive_interval: 2000,
  reconnect_timeout: 1000,
  topics: ["my_topic", "test"]
}
:my_client

16:45:39.514 [info]      :supervisor: {:local, :brod_sup}
    :started: [
  pid: #PID<0.310.0>,
  id: :my_client,
  mfargs: {:brod_client, :start_link,
   [[localhost: 9092], :my_client, [client_id: :my_client]]},
  restart_type: {:permanent, 10},
  shutdown: 5000,
  child_type: :worker
]
%{
  client_config: [client_id: :my_client],
  fetch_config: %{},
  group_config: [offset_commit_policy: :commit_to_kafka_v2],
  group_id: "group_1",
  hosts: [localhost: 9092],
  offset_commit_on_ack: true,
  offset_reset_policy: :latest,
  receive_interval: 2000,
  reconnect_timeout: 1000,
  topics: ["my_topic", "test"]
}
:my_client
%{
  client_config: [client_id: :my_client],
  fetch_config: %{},
  group_config: [offset_commit_policy: :commit_to_kafka_v2],
  group_id: "group_1",
  hosts: [localhost: 9092],
  offset_commit_on_ack: true,
  offset_reset_policy: :latest,
  receive_interval: 2000,
  reconnect_timeout: 1000,
  topics: ["my_topic", "test"]
}
:my_client
%{
  client_config: [client_id: :my_client],
  fetch_config: %{},
  group_config: [offset_commit_policy: :commit_to_kafka_v2],
  group_id: "group_1",
  hosts: [localhost: 9092],
  offset_commit_on_ack: true,
  offset_reset_policy: :latest,
  receive_interval: 2000,
  reconnect_timeout: 1000,
  topics: ["my_topic", "test"]
}
:my_client
%{
  client_config: [client_id: :my_client],
  fetch_config: %{},
  group_config: [offset_commit_policy: :commit_to_kafka_v2],
  group_id: "group_1",
  hosts: [localhost: 9092],
  offset_commit_on_ack: true,
  offset_reset_policy: :latest,
  receive_interval: 2000,
  reconnect_timeout: 1000,
  topics: ["my_topic", "test"]
}
:my_client
```

Thus, all the 5 producers are sharing the same TCP connection.

This is the config that I used:

```elixir
    Broadway.start_link(MyBroadway,
      name: MyBroadwayExample,
      producer: [
        module: {BroadwayKafka.Producer, [
          hosts: [localhost: 9092],
          group_id: "group_1",
          client_config: [
            client_id: :my_client
          ],
          topics: ["my_topic", "test"],
        ]},
        concurrency: 5
      ],
      processors: [
        default: [concurrency: 10]
      ]
    )
```

Reply:
I think the idea here is that client_id should be unique for each client, I think client_id in broadway configuration might be something like client_id_prefix (or a better name) and so client_id passed to connection is a concatenation :"#{client_id_prefix}#{ unique_string_by_client}" that makes it unique.

Reply:
You are absolutely right @v0idpwn ! I thought that it would be suffixed anyway with _0, _1 and so on... but that is done by the `Broadway.Topology` and not the producer here and that is why it receives it as a parameter.

It took me a while to find where that was being set... 

Anyway, perhaps we should simply use a `client_id_prefix`that would be suffixed with the `producer_name` anyway. It is not beautiful to that with atoms but it is, at least, only run once at boot time. I will fix it here.

Reply:
Updated! THanks @v0idpwn